File: LeetCode/2370. Longest Ideal Subsequence/solution.cpp
class Solution {
public:
    int longestIdealString(string s, int k) {
        vector<int>mx(26);
        for(char ch:s){
            ch-='a';
            int x=0;
            for(int i=ch-k;i<=ch+k;++i)
                if(i>=0&&i<26)
                    x=max(x,mx[i]);
            mx[ch]=x+1;
        }
        return *max_element(mx.begin(),mx.end());
    }
};

File: LeetCode/2370. Longest Ideal Subsequence/solution.py
class Solution:
    def longestIdealString(self, s: str, k: int) -> int:
        n = len(s)
        dp = [0]*26
        for i in range(n)[::-1]:
            ndp = [0]*26
            ch = ord(s[i])-ord('a')
            for j in range(26):
                ndp[j] = dp[j]
                if abs(j-ch) <= k:
                    ndp[j] = max(ndp[j], 1+dp[ch])
            dp = ndp
        return max(dp)